cns/ipam/interfaces.go

```diff
@@ -0,0 +1,11 @@
+package ipam
+
+import (
+	"github.com/Azure/azure-container-networking/cns"
+	cnstypes "github.com/Azure/azure-container-networking/cns/types"
+	"github.com/Azure/azure-container-networking/crd/nodenetworkconfig/api/v1alpha"
+)
+
+type StateReconciler interface {
```

Comment:
It's a little odd to create a new package only for this one interface that you need in one place. It's likely better to move this to `./cns/nodesubnet` and unexport it.

Reply:
This isn't specific to nodesubnet though. Other deployment modes also have an IPAM state reconciliation step. I could put it in ./cns/ if you think that's not too broad.

Reply:
Perhaps what Tim is trying to say is more that in Go interfaces are defined by the consumer, not the provider, to describe the behavior that an object must have to be _passed in to them_.  It usually doesn't make sense to have a public interface in a separate package like this. 

I don't see anything preventing this from being a private interface in the nodesubnet package, why does it matter that other modes have IPAM state reconciliation?

Reply:
I can make it private, but that will leave two different private interfaces that are exactly the same - there's[this one](https://github.com/Azure/azure-container-networking/blob/master/cns/service/main.go#L1126) as well as what we are now discussing. If we keep this exposed, we can remove the other private interface in my next PR that will actually finish the implementation of nodesubnet.

Reply:
The point of [this interface](https://github.com/Azure/azure-container-networking/blob/master/cns/service/main.go#L1127C2-L1127C159) is to abstract the inputs in to [this function](https://github.com/Azure/azure-container-networking/blob/master/cns/service/main.go#L1132). _Anything_ that satisfies that interface signature is an acceptable input into that function. That's the only purpose that this interface definition exists for. Your code in a different package should define and accept a local interface which describes the behavior that it needs. If that happens to be the same right now, that's fine.
This has some context on the antipattern that we're trying to prevent from developing here: https://medium.com/@cep21/preemptive-interface-anti-pattern-in-go-54c18ac0668a

Reply:
Also relevant from the Go proverbs: ["A little copying is better than a little dependency"](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=9m28s)

Reply:
Done.